### PR TITLE
Add possibility to configure trust policy for role

### PR DESCRIPTION
**Pull request by waky79**
##### Issue Type:
- Feature Pull Request
##### Plugin Name:

iam
##### Summary:

Ansible iam role cannot be used by any other service than EC2

This change will add possibility to add trust relationship policy for iam role.

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @waky79. @defionscode @seiffert please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by gregdek** [comment]
Thanks @waky79. @seiffert please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

File: cloud/amazon/iam.py [hunk 1]
```diff
@@ -548,6 +557,22 @@ def main():
...
+    if trust_policy != None:
+        try:
+            json_data = open(trust_policy, 'r')
+        except IOError:
+            try:
+                apdoc = json.dumps(trust_policy)
+            except Exception as e:
+                module.fail_json(msg=str(e) + '\n' + module.params.get('trust_policy'))
+        else:
+            apdoc = json.dumps(json.load(json_data))
```
Comment (juliedavila) [inline cloud/amazon/iam.py]: This fails to catch badly formatted json causing an uncapatured exception to be thrown

``` shell
Traceback (most recent call last):
  File "/Users/jdavila/.ansible/tmp/ansible-tmp-1457729969.51-237985966879139/iam", line 3057, in <module>
    main()
  File "/Users/jdavila/.ansible/tmp/ansible-tmp-1457729969.51-237985966879139/iam", line 573, in main
    apdoc = json.dumps(json.load(json_data))
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 291, in load
    **kw)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```

Def want to capture this and give a sane error message to the user.

**Comment by juliedavila** [comment]
Does not work with the file lookup plugin 

example:

``` yml
---
- name: test IAM
  hosts: localhost
  connection: local
  tasks:
    - name: Create IAM Role
      iam:
        iam_type: role
        name: testing2684
        region: us-east-1
        state: present
        trust_policy: "{{lookup('file', 'files/policy.json')}}"

```

It causes

``` xml
"msg": "BotoServerError: 400 Bad Request\n<ErrorResponse xmlns=\"https://iam.amazonaws.com/doc/2010-05-08/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>MalformedPolicyDocument</Code>\n    <Message>This policy contains invalid Json</Message>\n  </Error>\n  <RequestId>a512904c-e7cc-11e5-97aa-215274170993</RequestId>\n</ErrorResponse>
```

**Comment by juliedavila** [comment]
Passing in a json string also fails, I tried with and w/o quotes

example

``` yml
---
- name: test IAM
  hosts: localhost
  connection: local
  tasks:
    - name: Create IAM Role
      iam:
        iam_type: role
        name: testing2684
        region: us-east-1
        state: present
        trust_policy: '{"Version":"2008-10-17","Statement":[{"Effect":"Deny","Principal":{"Service":"ec2.amazonaws.com"},"Action":"sts:AssumeRole"}]}'

```

**Comment by waky79** [comment]
I will add those, will take some time. Off for the next three weeks :(

**Comment by juliedavila** [comment]
No problem, enjoy your vacation

**Comment by waky79** [comment]
@defionscode sorry, decided to look at it and pushed fixes. Tested with the set you gave + with standard file.

**Comment by juliedavila** [comment]
Oh geeze, @waky79 just realized I overlooked something. This should be a feature add to the `iam_policy` module and not to the `iam` module. Very sorry I didn't catch that sooner. It shouldn't be terribly difficult to port over the feature. 

This is likely the method that you'd want to use instead.

http://boto.cloudhackers.com/en/latest/ref/iam.html#boto.iam.connection.IAMConnection.update_assume_role_policy

**Comment by gregdek** [comment]
Thanks @waky79 for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

**Comment by gregdek** [comment]
Thanks @waky79 for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @waky79 for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@waky79 A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**